### DISCUSSION TITLE
Any method to remove stroke fill color in a jpg image?

### DISCUSSION THREAD

**Discussion by nissansz** [python-pillow/Pillow#8133] [Q&A] [answered]
![image](https://github.com/python-pillow/Pillow/assets/41010669/a20a3722-25e7-415c-9ab7-665a19fae005)

**Comment by radarhere** [comment, marked as answer]
Pillow can filter out specific colours and modify specific pixels, but as far as determining automatically which pixels are part of the stroke,

https://github.com/python-pillow/Pillow/issues/7376#issuecomment-1706317192
> Pillow doesn't have any functionality for text detection.

https://github.com/python-pillow/Pillow/issues/4572#issuecomment-618246247
> We don't currently ship any techniques for edge detection, so I'm not sure that this is the appropriate library for that. OpenCV or SciPy would probably be better choices.

**Comment by nissansz** [comment]
I realized that this effect is slightly different from stroke color.

It is more heavy shading.

Any method to draw such effect in pillow with text?









--------------------------------------------------------------------------------





------------------ 原始邮件 ------------------




 
 发件人: Andrew Murray  ***@***.***>
 

 
 发送时间: 2024-06-12 06:22:33
 

 
 收件人:python-pillow/Pillow ***@***.***> 

 
 抄送:nissanjp ***@***.***>,Author ***@***.***> 

 
 主题: Re: [python-pillow/Pillow] Any method to remove stroke fill color ina jpg image? (Discussion #8133)
 





 Pillow can filter out specific colours and modify specific pixels, but as far as determining automatically which pixels are part of the stroke,

 #7376 (comment)

 Pillow doesn't have any functionality for text detection.

 #4572 (comment)

 We don't currently ship any techniques for edge detection, so I'm not sure that this is the appropriate library for that. OpenCV or SciPy would probably be better choices.

**Reply by nissansz** [reply]
found a method, not sure whether it is good or not.
write text, blur it, 
then draw clean text on blurred image again

![image](https://github.com/python-pillow/Pillow/assets/41010669/e62ea7f3-2162-49e5-a17b-4198b4f771d7)

**Reply by radarhere** [reply]
Yes, that is a reasonable strategy.

**Reply by nissansz** [reply]
Any other suitable and better strategy?

**Reply by radarhere** [reply]
I don't know what you mean by 'better', but you can read at https://pillow.readthedocs.io/en/stable/reference/ImageFilter.html about different blurring methods.

**Reply by nissansz** [reply]
I mean faster, more natrual, etc.

**Reply by radarhere** [reply]
'more natural' is vague, and I don't know what would be faster because I don't know what method you're currently using.

**Reply by nissansz** [reply]
Below is what I used.

```python
  font0 = ImageFont.truetype(font_path, size=font_size-5)

    fontpath1 = r"F:\fonts\NotoOldShapeNormal.TTF"

    font1 = ImageFont.truetype(fontpath1, size=font_size-5)
    # font_family = ImageFont.FreeTypeFontFamily(font0, font1, font2)
    font_family = ImageFont.FreeTypeFontFamily(font0, font1)

    # font.set_variation_by_name("Black")

    im = Image.new("L", font_family.getbbox(text)[2:])
    draw = ImageDraw.Draw(im)
    # draw.text((0, 0), text, "#fff", font_family)
    draw.text((0, 0), text, font=font_family, fill="#fff", stroke_width=2, stroke_fill="#000")  # 彩色底色 配黑字

    # final = im
    # 反色
    im = Image.eval(im, lambda x: 255 - x)


    # Load random background image
    background = random_image_from_directory(background_dir)
    # background = background.resize((width, height))
    background = background.resize((im.width, im.height))

    # Initialize ImageDraw object
    draw = ImageDraw.Draw(background)

# 用反色写一次文字，模糊，然后再在模糊图上写一次清晰文字
    draw.text((0, 0), text, "#fff", font_family)
    # draw.text(position, text1, font=fnt, fill=color, stroke_width=strokewidth,stroke_fill=strokecolor)  # 彩色底色 配黑字
    # draw.text((0, 0), text, font=font_family, fill="#fff", stroke_width=2, stroke_fill="#000")  # 彩色底色 配黑字

    text_color = (128, 128, 128)  # 灰色 RGB 值
    # draw.text((5, 5), text, font=font_family, fill="#000", stroke_width=3, stroke_fill="#fff")  # 彩色底色 配黑字
    draw.text((5, 1), text, font=font_family, fill= text_color , stroke_width=3, stroke_fill="#fff")  # 彩色底色 配黑字

    # 添加阴影效果
    shadow = background .filter(ImageFilter.BLUR)
    # shadow = background .filter(ImageFilter.BoxBlur(7))
    # background .paste(shadow, (52, 52))

    draw = ImageDraw.Draw( shadow)
    draw.text((0, 0), text, "#fff", font_family)
    # draw.text(position, text1, font=fnt, fill=color, stroke_width=strokewidth,stroke_fill=strokecolor)  # 彩色底色 配黑字
    # draw.text((0, 0), text, font=font_family, fill="#fff", stroke_width=2, stroke_fill="#000")  # 彩色底色 配黑字
    draw.text((0, 0), text, font=font_family, fill="#fff")  # 彩色底色 配黑字
```

**Reply by radarhere** [reply]
Ok, so you've tried `ImageFilter.BLUR` and `ImageFilter.BoxBlur.` You could try `ImageFilter.GaussianBlur` and see if that is faster or not.

I don't know if your background is significantly larger than your text, but if it is, you could try drawing the text onto a smaller image first, blurring that, and then pasting it onto the final result. That may be faster.

**Reply by nissansz** [reply]
Do you mean crop is faster than resize to get a target size image for drawing?


original jpg background image size is 1280x720, how to  crop a text-line length subimage for draw?
Is below method fine?

```python
def random_crop_text_line(image_path, text_length):
    # Open the image
    img = Image.open(image_path)
    
    # Get the width and height of the image
    width, height = img.size
    
    # Define the maximum x-coordinate to crop from
    max_x = width - text_length
    
    # Randomly select the starting x-coordinate for cropping
    start_x = random.randint(0, max_x)
    
    # Crop the subimage with text-line length
    subimg = img.crop((start_x, 0, start_x + text_length, height))
    
    return subimg
```

**Reply by radarhere** [reply]
I meant that if your goal is only to blur the text, and not the background, then blurring a smaller image would be faster than blurring a larger image.

```python
from PIL import Image, ImageDraw, ImageFilter, ImageFont
import time
im = Image.new("RGB", (1000, 1000), "#f00")
font = ImageFont.load_default(48)

# Drawing the text and then blurring the whole image is slower
start = time.time()
d = ImageDraw.Draw(im)
d.text((10, 10), "text", fill="#000", font=font)
im = im.filter(ImageFilter.BLUR)
print(time.time() - start)

im = Image.new("RGB", (1000, 1000), "#f00")

# Blurring just the text is faster
start = time.time()
left, top, right, bottom = font.getbbox("text")
text_im = Image.new("RGBA", (right + 20, bottom + 20))
d = ImageDraw.Draw(text_im)
d.text((10, 10), "text", fill="#000", font=font)
text_im = text_im.filter(ImageFilter.BLUR)
text_im.save("out.png")
im.paste(text_im, mask=text_im)
print(time.time() - start)
```

**Reply by nissansz** [reply]
So, I need 3 layers?
1. crop a colorful background.
2. blur a text layer
3. paste blue layer on background
4. draw clear text on step 3?

**Reply by radarhere** [reply]
I don't know what you are trying to achieve or what steps you need to get there - this is the first time you've talked about anything blue in this conversation.

I'm just saying that blurring only the text would be faster than blurring a large image.

**Reply by nissansz** [reply]
Ok. Thank you.
